sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala

```diff
@@ -56,3 +57,9 @@ private[sql] abstract class BucketedOutputWriterFactory extends OutputWriterFact
       context: TaskAttemptContext): OutputWriter =
     throw new UnsupportedOperationException("use bucket version")
 }
+
+private[sql] object BucketingUtils {
```

Comment:
+1 

Structurally it is better for bucketing to be a user of a general hash expression, rather than the other way around.